pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```diff
@@ -24,307 +24,480 @@
+        for (int i = startIndex; i < endIndex; i++) {
+          IntArrayList rawKey = new IntArrayList(_numGroupByColumns);
+          rawKey.size(_numGroupByColumns);
+          int[] rawKeyArray = rawKey.elements();
+          for (int j = 0; j < _numGroupByColumns; j++) {
+            rawKeyArray[j] = _reusableSingleDictIds[j][i];
           }
+          docIdToGroupKey[outIndex++] = updateRawKeyToGroupKeyMapping(rawKey);
         }
-      }
+        break;
+      default:
+        throw new RuntimeException("Unsupported storage type.");
     }
-
-    int numGroupKeys = groupKeysList.size();
-    int[] groupKeys = new int[numGroupKeys];
-    for (int i = 0; i < numGroupKeys; i++) {
-      int groupKey = updateRawKeyToGroupKeyMapping(groupKeysList.getLong(i));
-      groupKeys[i] = groupKey;
-    }
-    return groupKeys;
   }
 
   /**
-   * Given a docIdSet, return a map containing array of group by keys for each docId.
-   *
-   * @param docIdSet
-   * @param length
-   * @param docIdToGroupKeys
+   * {@inheritDoc}
    */
   @Override
   public void generateKeysForDocIdSet(int[] docIdSet, int startIndex, int length, int[][] docIdToGroupKeys) {
+    // Fetch all dictionary ids according to the document id set for all single value group-by columns.
     for (int i = 0; i < _numGroupByColumns; i++) {
       if (_isSingleValueGroupByColumn[i]) {
         _singleBlockValSets[i].readIntValues(docIdSet, startIndex, length, _reusableSingleDictIds[i], 0);
       }
     }
+
+    // Calculate the group keys(int[]) and store it into the result buffer.
     int outIndex = 0;
     int endIndex = startIndex + length;
-    for (int i = startIndex; i < endIndex; i++) {
-      docIdToGroupKeys[outIndex++] = generateKeysForDocId(outIndex, docIdSet[i]);
+    switch (_storageType) {
+      case ARRAY_BASED:
+        for (int i = startIndex; i < endIndex; i++) {
+          docIdToGroupKeys[outIndex++] = generateKeysForDocIdArrayBased(outIndex, docIdSet[i]);
+        }
+        break;
+      case LONG_MAP_BASED:
+        for (int i = startIndex; i < endIndex; i++) {
+          docIdToGroupKeys[outIndex++] = generateKeysForDocIdLongMapBased(outIndex, docIdSet[i]);
+        }
+        break;
+      case ARRAY_MAP_BASED:
+        for (int i = startIndex; i < endIndex; i++) {
+          docIdToGroupKeys[outIndex++] = generateKeysForDocIdArrayMapBased(outIndex, docIdSet[i]);
+        }
+        break;
+      default:
+        throw new RuntimeException("Unsupported storage type.");
     }
   }
 
   /**
-   * With an integer raw key, convert group key from dictId based to string based, using actually values corresponding
-   * to dictionary id's.
-   *
-   * @param rawKey
-   * @return
+   * {@inheritDoc}
    */
-  private String rawKeyToStringGroupKey(int rawKey) {
-    // Special case one group by column for performance.
-    if (_numGroupByColumns == 1) {
-      return _dictionaries[0].get(rawKey).toString();
+  @Override
+  public int getMaxNumGroupKeys() {
```

Comment:
Method name is a bit confusing, may be rename it getGroupByCardinalityProduct(), or getMaxPossibleNumGroupKeys()?